File: src/offer8.Next_Node_of_Binary_Tree/next_node_of_binary_tree.cpp
/*
struct TreeLinkNode {
    int val;
    struct TreeLinkNode *left;
    struct TreeLinkNode *right;
    struct TreeLinkNode *next;
    TreeLinkNode(int x) :val(x), left(NULL), right(NULL), next(NULL) {
        
    }
};
*/
class Solution {
public:
    TreeLinkNode* GetNext(TreeLinkNode* pNode)
    {
        if(pNode == nullptr) return nullptr;
        if(pNode->right != nullptr){
            pNode = pNode->right;
            while(pNode->left != nullptr){
                pNode = pNode->left;
            }
            return pNode;
        }
        while(pNode->next != nullptr){
            TreeLinkNode* nroot = pNode->next;
            if(nroot->left == pNode){
                return nroot;
            }
            pNode = pNode->next;
        }
        return nullptr;
    }
};

File: src/offer08.Next_Node_of_Binary_Tree/next_node_of_binary_tree.java
//有右子树
//无右子数
    //当前是最左子孩子，返回他的父节点
    //当前不是最左子孩子，向上找
/*
public class TreeLinkNode {
    int val;
    TreeLinkNode left = null;
    TreeLinkNode right = null;
    TreeLinkNode next = null;

    TreeLinkNode(int val) {
        this.val = val;
    }
}
*/
public class Solution {
    public TreeLinkNode GetNext(TreeLinkNode pNode)
    {
        if(pNode == null)
        {
            return null;
        }
        if(pNode.right != null){
            pNode = pNode.right;
            while(pNode.left != null){
                pNode = pNode.left;
            }
            return pNode;
        }
        while(pNode.next != null){
            TreeLinkNode root = pNode.next;
            if(root.left == pNode){
                return root;
            }
            pNode = pNode.next;
        }
        return null;
    }
}